### ISSUE TITLE
Clarify how struct.new_rtt and ref.cast are related

### ISSUE THREAD

**Issue by taralx** [WebAssembly/gc#49]
Reading the MVP, it is not clear what the rtt argument to `struct.new_rtt` does. It's not referenced anywhere else, most surprisingly not in `ref.cast`.

Can someone help explain?

**Comment by rossberg** [comment]
I added a couple of minor clarifications, in particular, by linking to the #values section that defines how reference values carry runtime types. Please let me know if the relation is still unclear.

**Comment by taralx** [comment]
Thanks! If I read this correctly then, the rtt passed to new constrains how far down one can cast, with a limit of the actual type?

**Comment by taralx** [comment]
While I'm in here asking rtt questions - why does rtt.new take a supertype rtt? It looks like rtt.anyref will always suffice.

**Comment by rossberg** [comment]
The RTTs essentially implement a nominal "single inheritance" subtyping hierarchy, thereby enabling the usual efficient methods for (almost) constant time runtime down casts. You can only case a value to its most specific type or one of the supertypes in this RTT hierarchy.

**Comment by taralx** [comment]
Understood, but why do you need a supertype to make a new rtt? (rtt.anyref) is sufficient in every case AFAICT.

**Comment by taralx** [comment]
More specifically: Can you give an example where one would pass something other than rtt.anyref to rtt.new?

**Comment by rossberg** [comment]
Consider your favourite inheritance-based OO language:
```
class A
class B < A
class C < B

func f(x : A) { downcast<B>(x) }
f(new C)
```
To compile this, you'd have to mirror the class hierarchy as an RTT hierarchy.

**Comment by taralx** [comment]
What wrong with `(rtt.anyref) (rtt.new anyref $cls-B) (ref.cast anyref $cls-B)`? No hierarchy required.

**Comment by rossberg** [comment]
Not sure I understand your suggestion, but all the following cases need to work, of course:
```
func A2B(x : A) { downcast<B>(x) }
func A2C(x : A) { downcast<C>(x) }
func B2C(x : B) { downcast<C>(x) }
A2B(new B)
A2B(new C)
A2C(new C)
B2C(new C)
```

**Comment by Horcrux7** [comment]
Is there already a reference implementation with rtt support  with which we can play to understand this better?

Why is there no struct.new_rtt_default or struct.new_default_rtt? My understand is that struct.new_rtt bind the type information to the created instance. This is the difference to struct.new. Then there should be the same to struct.new_default.

The struct.new and struct.new_default create an instance of a type without to bind the type information to the instance. How can a GC work without the type information in every instance? My understand is that the GC must know which fields of the instance contains references to following it. But of course I have never write a GC and I think to simple.

**Comment by taralx** [comment]
> Not sure I understand your suggestion, but all the following cases need to work, of course:
> 
> ```
> func A2B(x : A) { downcast<B>(x) }
> func A2C(x : A) { downcast<C>(x) }
> func B2C(x : B) { downcast<C>(x) }
> A2B(new B)
> A2B(new C)
> A2C(new C)
> B2C(new C)
> ```

And all of those work without any need to pass something other than (rtt.anyref) to (ref.cast). So my question is why does ref.cast want an rtt at all?

**Comment by rossberg** [comment]
@Horcrux7, no, there is no reference interpreter yet, as we are still discussing the basic design.

The `new` form without RTT is just a shorthand passing `rtt.anyref`; `new_default` is a shorthand of the shorthand passing default values (where possible). We could also have the other shorthand as you suggest, but it may be less relevant. We could also remove one or both of the others. This surely is something to revisit with more producer feedback.

RTT information is not the same as layout information for the GC. An implementation may combine these, but depending on implementation details, the representation of a heap value without RTT could potentially be simpler and use less space.

**Comment by rossberg** [comment]
@taralx, how would casting from A to B succeed for a value that is actually a C without the dynamic information that that B is a supertype of C?
